rust/src/origin.rs

```diff
@@ -6,56 +6,204 @@
  * SPDX-License-Identifier: Apache-2.0 OR MIT
  */
 
-use anyhow::{bail, Result};
```

Comment:
Note for this code it's not really useful to review the diff because the new code is different in purpose and structure from the old one, so just look at the new code.